ssh/ssh_gocrypto.go

```diff
@@ -234,10 +256,250 @@ func (c *goCryptoCommand) StderrPipe() (io.ReadCloser, io.Writer, error) {
+		return terminal.NewTerminal(os.Stdin, ""), cleanup, nil
+	}
+	return nil, nil, nil
+}
+
+// checkHostKey checks the given (hostname, address, public key) tuple
+// against the local known-hosts database, if it exists, and returns a
+// boolean indicating whether a match was found, and any errors encountered.
+func checkHostKey(
+	hostname string,
+	remote net.Addr,
+	key ssh.PublicKey,
+	knownHostsFile string,
+	printError func(string) error,
+) (bool, error) {
+	// NOTE(axw) the knownhosts code is incomplete, but enough for
+	// our limited use cases. We do not support parsing a known_hosts
+	// file managed by OpenSSH (due to hashed hosts, etc.), but that
+	// is OK since this client exists only to support systems that
+	// do not have access to OpenSSH.
+	callback, err := knownhosts.New(knownHostsFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// The known_hosts file does not exist.
+			return false, nil
+		}
+		return false, errors.Trace(err)
+	}
+	err = callback(hostname, remote, key)
+	switch err := err.(type) {
+	case nil:
+		// Known host with matching key.
+		return true, nil
+	case *knownhosts.KeyError:
+		if len(err.Want) == 0 {
+			// Unknown host.
+			return false, nil
+		}
+		head := fmt.Sprintf(`
+@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
+@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
+Someone could be eavesdropping on you right now (man-in-the-middle attack)!
+It is also possible that a host key has just been changed.
+The fingerprint for the %s key sent by the remote host is
+%s.
+Please contact your system administrator.
+Add correct host key in %s to get rid of this message.
+`[1:], key.Type(), ssh.FingerprintSHA256(key), knownHostsFile)
+
+		var typeKey *knownhosts.KnownKey
+		for i, knownKey := range err.Want {
+			if knownKey.Key.Type() == key.Type() {
+				typeKey = &err.Want[i]
+			}
+		}
+
+		var tail string
+		if typeKey != nil {
+			tail = fmt.Sprintf(
+				"Offending %s key in %s:%d",
+				typeKey.Key.Type(),
+				typeKey.Filename,
+				typeKey.Line,
+			)
+		} else {
+			tail = "Host was previously using different host key algorithms:"
+			for _, knownKey := range err.Want {
+				tail += fmt.Sprintf(
+					"\n - %s key in %s:%d",
+					knownKey.Key.Type(),
+					knownKey.Filename,
+					knownKey.Line,
+				)
+			}
+		}
+		if err := printError(head + tail); err != nil {
+			// Not being able to display the warning
+			// should be considered fatal.
+			return false, errors.Annotate(
+				err, "failed to print host key mismatch warning",
+			)
+		}
+	}
+	return false, errors.Trace(err)
+}
+
 func splitUserHost(s string) (user, host string) {
 	userHost := strings.SplitN(s, "@", 2)
 	if len(userHost) == 2 {
 		return userHost[0], userHost[1]
 	}
 	return "", userHost[0]
 }
+
+var (
+	goCryptoKnownHostsMutex sync.Mutex
+	goCryptoKnownHostsFile  string
```

Comment:
Maybe if we did things over, but we have several globals already, including the client itself (i.e. ssh.DefaultClient). It's not worth the effort to try and change it around right now.